8320275: assert(_chunk->bitmap().at(index)) failed: Bit not set at index

pchilano opened the pull request:
Please review the following fix. The assert fails while verifying the top frame of the stackChunk before returning from a thaw call. The stackChunk is in gc mode but we found a narrow oop for this c2 compiled frame that doesn't have its corresponding bit set. This is because while thawing its callee we cleared the bitmap range associated with the argument area, but this narrow oop happens to land at the very last stack slot of that region.
Loom code assumes the size of the argument area is always a multiple of 2 stack slots, as SharedRuntime::java_calling_convention() shows. But c2 doesn't seem to follow this convention and, knowing the last passed argument only takes one stack slot, it's using the remaining space to store a narrow oop for the caller. There are more details about the specific crash in JBS.

The initial proposed fix is to just restrict the range of the bitmap we clear by excluding the last stack slot of the argument area, since passed oops are always word aligned. I've also experimented with a patch where I changed SharedRuntime::java_calling_convention() and Fingerprinter::do_type_calling_convention() to not round up the number of stack slots used, and then changed the callers to use a round up value or not depending on the needs [1]. I wasn't convinced it was worthy given we only care about this difference in this Loom code, but I don't mind going with that fix instead. The 3rd alternative would be to just change c2 to not use this stack slot and start spilling at a word aligned offset from the sp.

I run the patch with the failing test and verified the crash doesn't reproduce anymore. I've also run this patch through loom tiers1-5. 

Thanks,
Patricio

[1] https://github.com/pchilano/jdk/commit/42ae9269b28beb6f36c502182116545b680e418f

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320275](https://bugs.openjdk.org/browse/JDK-8320275): assert(_chunk-&gt;bitmap().at(index)) failed: Bit not set at index (**Bug** - P3)


### Reviewers
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)
 * [Frederic Parain](https://openjdk.org/census#fparain) (@fparain - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16837/head:pull/16837` \
`$ git checkout pull/16837`

Update a local copy of the PR: \
`$ git checkout pull/16837` \
`$ git pull https://git.openjdk.org/jdk.git pull/16837/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16837`

View PR using the GUI difftool: \
`$ git pr show -t 16837`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16837.diff">https://git.openjdk.org/jdk/pull/16837.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16837#issuecomment-1828873127)

pchilano commented:
Running some extra tests I see the callee can use the argument area to store data that is different from the one passed. This is actually something @fparain told me some time ago. So this simpler solution won't do. Before applying https://github.com/pchilano/jdk/commit/42ae9269b28beb6f36c502182116545b680e418f instead, @dean-long  how about if we just prevent c2 from using this stack slot for the caller?

dean-long commented:
I don't really like the use of `address` for the pointer that might be either `oop` or `narrowOop`, because the bitmap can't really support an arbitrary `char*` address.  I think it would be better to cleanup methods that take `intptr_t*` and instead use `template <typename OopT>` like `bit_index_for`.
Also, I would prefer removing the round up in `java_calling_convention`, because the current patch fixes a subtle bug with an even subtler work-around, but let's see what other reviews think.

pchilano commented:
>I don't really like the use of address for the pointer that might be either oop or narrowOop, because the bitmap can't really support an arbitrary char* address. I think it would be better to cleanup methods that take intptr_t* and instead use template <typename OopT> like bit_index_for.
>
The thing is that we would need to check before calling `clear_bitmap_bits()` if `UseCompressedOops` is set to use oop or narrowOop, which makes the code uglier in my opinion. This is a private method anyways and is not meant to be used outside this two callers in thaw code. But I added asserts in `clear_bitmap_bits()` to verify `start` and `end` have the expected alignment. I also added an is_aligned() assert in `stackChunkOopDesc::bit_index_for()` just for extra caution. While adding this last assert I actually realized I had a bug in my original experiment branch because I missed to align_down `end` when `UseCompressedOops` is not set, so thanks for pointing this out.

pchilano commented:
I tested the last version in mach5 loom-tiers[1-5] and with the failing test. I'll keep running more rounds though since issues with this code are highly intermittent.

pchilano commented on src/hotspot/share/runtime/continuationFreezeThaw.cpp:
```diff
@@ -2166,13 +2166,17 @@ inline void ThawBase::patch(frame& f, const frame& caller, bool bottom) {
   // we need to clear the bits that correspond to arguments as they reside in the caller frame
   // or they will keep objects that are otherwise unreachable alive
-  log_develop_trace(continuations)("clearing bitmap for " INTPTR_FORMAT " - " INTPTR_FORMAT, p2i(start), p2i(start+range));
+  address effective_end = UseCompressedOops ? end : align_down(end, wordSize);
```
Because `end` is not necessarily word aligned anymore the pointer arithmetic we do in bit_index_for() would be UB, since `p` can point to the middle of an oop (in practice we would probably not see any issue because that's implemented as a substraction and then an arithmetic shift right which will round down the result). So we need to align `end` down if UseCompressedOops is not set. That last half word part should not contain an oop anyways so the assert is to verify that. I added a comment, please take a look.

dean-long commented on src/hotspot/share/runtime/continuationFreezeThaw.cpp:
```diff
@@ -2298,7 +2309,10 @@ void ThawBase::recurse_thaw_compiled_frame(const frame& hf, frame& caller, int n
   } else if (_cont.tail()->has_bitmap() && added_argsize > 0) {
-    clear_bitmap_bits(heap_frame_top + ContinuationHelper::CompiledFrame::size(hf) + frame::metadata_words_at_top, added_argsize);
+    address start = (address)(heap_frame_top + ContinuationHelper::CompiledFrame::size(hf) + frame::metadata_words_at_top);
+    int stack_args_slots = f.cb()->as_compiled_method()->method()->num_stack_arg_slots(false /* rounded */);
```
It would be nice if we could trust the `added_argsize` value here, but that would require more changes to where rounding is done.

pchilano commented:
Thanks for the reviews @dean-long and @fparain!
I want to give it a couple more rounds of testing in the upper tiers. Since I'll be out on vacation after today that means I'll integrate this once I'm back.

pchilano commented:
/integrate
